WIP: [#26] Allow snake to eat (#27)
codecurious-bln/elixir_snake: docs/tutorial/06-allow-snake-to-eat.md

### 6. Allow snake to eat

We have the food, but our snake can't eat it yet, and we don't want to starve our little friend.

The goal of this chapter is to allow the snake to eat the food and grow, but let's first quickly recap how it works. The snake eats the food when it overlaps it with its head, as soon the snake eats it, it will grow of one unit (a new tile will be appended to its body) and a new food pellet will be placed in the snake world.

In other words, we need to check if the snake's head has the same coordinates of the pellet after every movement. The function `move_snake/1` looks the natural place where to execute this check.

```elixir
defp move_snake(%{snake: snake} = state) do
  %{body: body, size: size, direction: direction} = snake

  # New head's position
  [head | _] = body
  new_head = move(state, head, direction)

  # Place a new head on the tile that we want to move to
  # and remove the last tile from the snake tail
  new_body = List.delete_at([new_head | body], -1)

  state
  |> put_in([:objects, :snake, :body], new_body)

>   |> maybe_eat_food(new_head)

Author: 👍  Should we drop the `maybe`? 

```elixir
state
|> put_in([:objects, :snake, :body], new_body)
|> eat_pellet(new_head)
```

What do. you think?

>   |> maybe_eat_food(new_head)

Reviewer: For now not necessary in my opinion. I think the names are a bit off here - e.g. `move_snake` does more than moving the snake - but in general this is still fine I'd say.

The "problem" with `maybe_` that the post tasks about also does not really fit our case I'd say.

>   |> maybe_eat_food(new_head)

Author: > For now not necessary in my opinion. I think the names are a bit off here - e.g. move_snake does more than moving the snake - but in general this is still fine I'd say.

Good point, what about `advance_snake` ?

> The "problem" with `maybe_` that the post tasks about also does not really fit our case I'd say.

Sorry, but I didn't get your comment. Do you mean that the function name does not match the current use case since it does more stuff like placing the new pellet? 🤔 
I see the point, but unfortunately I cannot came with a better name.
I kinda like that the `grow_snake` and `place_pellet` functions are grouped together since they are related. What about `maybe_grow_snake` or `check_xxx` dunno 😞

>   |> maybe_eat_food(new_head)

Reviewer: I guess whether it's "move" or "advance" does not really matter. As said, both is fine here in my opinion.

My point with the `maybe_` part was: the blog post has a more obvious scenario where one can get rid of the "maybe" part. Our case is different, we want **always** run this check.

Therefore: let's stick with the current naming - sorry for the confusion 🙈

end
```

👆Let's add it at the end of the state update pipeline, the pipe operator `|>` comes in handy here 💚

Our new function `maybe_eat_food/2` receives:

- The current state as 1st argument
- The snake's head coordinates as 2nd argument

Then, in the case the snake's head overlaps the food pellet, the function will take care of:

- Grow the snake body of one unit
- Place a new food pellet in the snake world

or otherwise, just return the current state without any changes.

Let's draft out the `maybe_eat_pellet/2` function and implement it step by step.

```elixir

> def maybe_eat_pellet(state = %{pellet: pellet}, snake_head) do

Author: Yup, good catch!

  if (pellet == snake_head)
    state
    |> grow_snake()
    |> place_pellet()
  else
    state
  end
end
```

> Growing the snake body can be tricky, let's explore all our alternatives.

Reviewer: ```suggestion
Growing the snake's body can be tricky. Let's explore some possibilities:
```

> Prepending a new tile to the snake head is not a feasible solution because it will potentially lead to unexpected outcome like its dead :skull:.

Reviewer: ```suggestion
Prepending a new tile before the snake's head is not feasible. It could lead to unexpected outcome: the snake hits its tail and dies ☠️
```

> The most natural approach is to append a new tile at the end of the snake body, but how exactly? We can not simply add a new tile at the end like: `Enum.concat(body, [{x, y}])` since we don't have any information of the tail direction but only of its head. In other words, we can't infer the coordinates (`{x, y}`) of the tile to append. The only way to safely grow our snake it's to preserve its body when the snake ate the pellet. We could set a boolean flag `has_eaten` in the state and in the next game tick, don't delete the last snake's body tile when this flag is true 🤓.

Reviewer: ```suggestion
The most natural approach is appending a new tile to the end of the body. But how? We can't append a tile with `Enum.concat(body, [{x, y}])`. We can't infer the coordinates (`{x, y}`) for the new tile, since we don't know the tail's direction. We only know where the head is moving.

The only way to safely grow our snake its to preserve its body when it ate the pellet. We can set a boolean flag `has_eaten` in the state and on the next tick not delete the last body tile if is set to `true` 🤓. This will naturally grow our snake by one tile.
```

Remember, we set a timer at the beginning in out `init/1` function that periodically sends a message, which is intercepted by our `handle_info/2`, which in turn calls the `move_snake/1` function. That's our game tick.

```elixir
defp move_snake(%{snake: snake} = state) do
  %{body: body, size: size, direction: direction} = snake

  # New head's position
  [head | _] = body
  new_head = move(state, head, direction)

  # Place a new head on the tile that we want to move to
  # and remove the last tile from the snake tail
  new_body = List.delete_at([new_head | body], -1)

  state
  |> put_in([:snake, :body], new_body)

>   |> maybe_eat_food(new_head)

Reviewer: This does not match the name of the later implemented function 🤔

Author: ops.

end

def maybe_eat_pellet(state = %{pellet: pellet}, snake_head) do
  if (pellet == snake_head)
    state
    |> grow_snake()
    |> place_pellet()
  else
    state
  end
end

def grow_snake(state = %{%{snake: %{size: size}}) do
  put_in(state, [:snake, :has_eaten], true)
end

def place_pellet(state = %{width: width, height: height, snake: %{body: snake_body}}) do
  pellet_coords = {
    Enum.random(0..(width - 1)),
    Enum.random(0..(height - 1))
  }

  if pellet_coords in snake_body do
    place_pellet(state)
  else
    put_in(state, [:objects, :pellet], pellet_coords)
  end
end
```

> Let's take a look to these two new functions:

Reviewer: Same here: I'd move this paragraph on top of the code snipped and introduce it more like:

> let's add two new functions..

And describe what they should do.

- `grow_snake/1` simply sets the `:has_eatan` flag to true in the state

> - `place_pellet/1` computes a new pair of coordinates for the food, if the new value matches any tile in the snake's body, it recursively generate a new position until it does not overlap the snake

Reviewer: ```suggestion
- `place_pellet/1` computes a new pair of coordinates for the food. If the new value matches any tile in the snake's body, it recursively generate a new position until it does not overlap any more.
```

We still need to update the `move_snake/1` function to use the `:has_eatan` flag.

```elixir
defp move_snake(%{snake: snake} = state) do
  %{body: body, direction: direction} = snake

  # New head's position
  [head | _] = body
  new_head = move(state, head, direction)

  # Place a new head on the tile that we want to move to
  # and remove the last tile from the snake tail if it has not eaten any pellet
  new_body = [new_head | body]
  new_body = if snake.has_eaten, do: new_body, else: List.delete_at(new_body, -1)

  state
  |> put_in([:snake, :body], new_body)
  |> maybe_eat_food(new_head)
end
```

> Like that, when the flag `:has_eatan` is true, the last tile from the snake's body is not removed anymore. This is the trick that allows us to grow the snake 💪

Reviewer: ```suggestion
When `:has_eaten` is true, the last tile from the snake's body is **not** removed. This is the trick that allows us to grow the snake 💪
```

And now let's run the game and see how if our snake grows when eating.

    $ mix scenic.run

> Oh snapp! Our snake is growing infinitely!! We forgot to reset the `:has_eaten` flag 🙈

Reviewer: Nice one 💚

Reviewer 2: this was really cool!

```elixir
defp move_snake(%{snake: snake} = state) do
  %{body: body, direction: direction} = snake

  # New head's position
  [head | _] = body
  new_head = move(state, head, direction)

  # Place a new head on the tile that we want to move to
  # and remove the last tile from the snake tail if it has not eaten any pellet
  new_body = [new_head | body]
  new_body = if snake.has_eaten, do: new_body, else: List.delete_at(new_body, -1)

  state
  |> put_in([:snake, :body], new_body)

>   |> put_in([::snake, :has_eaten], false) # Reset the `:has_eaten` flag before the next check

Reviewer: ```suggestion
  |> put_in([:snake, :has_eaten], false) # Reset the `:has_eaten` flag before the next check
```
